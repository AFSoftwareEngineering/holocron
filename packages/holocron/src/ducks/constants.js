/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

export const HOLOCRON_STORE_KEY = 'holocron';
export const MODULES_STORE_KEY = 'modules';
export const REGISTRY_MODULE_BLOCKED_KEY = 'blockedModules';
export const REGISTRY_MODULE_REDUCERS_KEY = 'reducers';
export const REGISTRY_MODULE_MAP_KEY = 'moduleMap';

export const HOLOCRON_KEY = '@@holocron';

export const REGISTER_MODULE_REDUCER = `${HOLOCRON_KEY}/REGISTER_MODULE_REDUCER`;
export const MODULE_LOADED = `${HOLOCRON_KEY}/MODULE_LOADED`;
export const MODULE_LOAD_FAILED = `${HOLOCRON_KEY}/MODULE_LOAD_FAILED`;
export const MODULE_LOADING = `${HOLOCRON_KEY}/MODULE_LOADING`;
export const MODULE_REDUCER_ADDED = `${HOLOCRON_KEY}/MODULE_REDUCER_ADDED`;
export const INIT_MODULE_STATE = `${HOLOCRON_KEY}/INIT_MODULE_STATE`;
export const REGISTER_MODULE = `${HOLOCRON_KEY}/REGISTER_MODULE`;
export const BLOCK_MODULE = `${HOLOCRON_KEY}/BLOCK_MODULE`;
export const SET_MODULE_MAP = `${HOLOCRON_KEY}/SET_MODULE_MAP`;
export const SET_MODULE_REDUCER = `${HOLOCRON_KEY}/SET_MODULE_REDUCER`;
export const RESET_MODULES_AND_MAP = `${HOLOCRON_KEY}/RESET_MODULES_AND_MAP`;

export const REDUCER_KEY = '@@holocron-module-reducer';
export const LOAD_KEY = '@@holocron-module-load-action';
