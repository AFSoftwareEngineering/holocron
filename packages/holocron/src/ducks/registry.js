/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { fromJS, Map as iMap, List as iList } from 'immutable';

import {
  REGISTRY_MODULES_KEY,
  REGISTRY_MODULE_MAP_KEY,
  REGISTRY_MODULE_BLOCKED_KEY,
  REGISTER_MODULE,
  BLOCK_MODULE,
  SET_MODULE_MAP,
  RESET_MODULES_AND_MAP,
} from './constants';

const initialState = iMap({
  [REGISTRY_MODULES_KEY]: iMap({}),
  [REGISTRY_MODULE_MAP_KEY]: iMap({ modules: iMap({}) }),
  [REGISTRY_MODULE_BLOCKED_KEY]: iList([]),
});

export function createInitialState(moduleMap, modules, blocked) {
  return fromJS({
    [REGISTRY_MODULES_KEY]: fromJS(modules || {}),
    [REGISTRY_MODULE_MAP_KEY]: fromJS(moduleMap || {}),
    [REGISTRY_MODULE_BLOCKED_KEY]: iList(blocked),
  });
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case RESET_MODULES_AND_MAP: {
      const { newModuleMap, newModules } = action;
      return state
        .update(REGISTRY_MODULES_KEY, iMap(),
          (modules) => modules.clear().merge(fromJS(newModules || {})))
        .update(REGISTRY_MODULE_MAP_KEY, iMap(),
          (moduleMap) => moduleMap.clear().merge(fromJS(newModuleMap || {})));
    }
    case BLOCK_MODULE: {
      const { moduleUrl } = action;
      return state
        .update(REGISTRY_MODULE_BLOCKED_KEY, iList(),
          (blockList) => blockList.push(moduleUrl));
    }
    case SET_MODULE_MAP: {
      const { newModuleMap } = action;
      return state
        .update(REGISTRY_MODULE_MAP_KEY, iMap(),
          (moduleMap) => moduleMap.clear().merge(fromJS(newModuleMap)));
    }
    case REGISTER_MODULE: {
      const { moduleName, module } = action;
      return state
        .update(REGISTRY_MODULES_KEY, iMap(),
          (modules) => modules.set(moduleName, module));
    }
    default:
      return state;
  }
}

export function registerModule(moduleName, module) {
  return {
    type: REGISTER_MODULE,
    moduleName,
    module,
  };
}

export function blockModule(moduleUrl) {
  return {
    type: BLOCK_MODULE,
    moduleUrl,
  };
}

export function setModuleMap(newModuleMap) {
  return {
    type: SET_MODULE_MAP,
    newModuleMap,
  };
}

export function resetModuleRegistry(newModules, newModuleMap) {
  return {
    type: RESET_MODULES_AND_MAP,
    newModules,
    newModuleMap,
  };
}

export function isModuleBlocked(moduleUrl) {
  return (state) => state.hasIn([REGISTRY_MODULE_BLOCKED_KEY, moduleUrl]);
}

export function getModule(moduleName) {
  return (state) => state.getIn([REGISTRY_MODULES_KEY, moduleName]);
}

export function getModules() {
  return (state) => state.getIn([REGISTRY_MODULES_KEY]);
}

export function getModuleMap() {
  return (state) => state.getIn([REGISTRY_MODULE_MAP_KEY]);
}

export function getBlockedModules() {
  return (state) => state.getIn([REGISTRY_MODULE_BLOCKED_KEY]);
}
