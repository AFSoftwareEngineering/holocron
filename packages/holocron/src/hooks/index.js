export {
  default as useHolocron,
  useHolocronContext,
  Holocron,
  HolocronContext,
} from './useHolocron';
