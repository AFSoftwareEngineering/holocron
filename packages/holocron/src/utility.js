/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';

import {
  REDUCER_KEY,
  LOAD_KEY,
  INIT_MODULE_STATE,
} from './ducks/constants';

export function isBrowser() {
  return !!global.BROWSER;
}

export function getGlobalInitialState() {
  // eslint-disable-next-line no-underscore-dangle
  return global.__INITIAL_STATE__;
}

export function getModuleDisplayName(name) {
  return `HolocronModule(${name})`;
}

export function getModuleName(Module) {
  return Module ? Module.displayName || Module.name : 'module';
}

export const getInitialState = (Module, props) => props.initialState
  || Module.holocron.initialState
  || (
    typeof props.reducer === 'function'
    && props.reducer(undefined, { type: INIT_MODULE_STATE })
  )
  || (
    typeof Module.holocron.reducer === 'function'
    && Module.holocron.reducer(undefined, { type: INIT_MODULE_STATE })
  );

export const getShouldModuleReload = (Module) => (
  Module.holocron.shouldModuleReload
    ? (prevProps, nextProps) => !Module.holocron.shouldModuleReload(prevProps, nextProps)
    : undefined);

export const getMergeProps = (Module, onMergeProps, defaultProps) => {
  if (Module && typeof Module.holocron.mergeProps === 'function') {
    return onMergeProps(Module.holocron.mergeProps) || defaultProps;
  }
  return defaultProps || undefined;
};

export const createLoad = (load) => (props) => (dispatch) => dispatch(load(props));
export const createLoadModuleData = (
  loadModuleData,
  Module
) => (props) => (dispatch, getState, { fetchClient }) => loadModuleData({
  store: { dispatch, getState },
  fetchClient,
  ownProps: props,
  module: Module,
});
export const getLoader = (Module, props) => {
  const getLoad = () => props.load || Module.holocron.load;
  const getLoadModuleData = () => props.loadModuleData
    || Module.holocron.loadModuleData
    || Module.loadModuleData;
  let loader = () => undefined;
  if (getLoad()) {
    console.warn(`The Holocron Config in '${Module.displayName}' is using the 'load' function which has been deprecated. Please use 'loadModuleData' instead.`);
    // TODO: deprecation (load) - remove in next major version
    loader = createLoad(getLoad());
  } else if (getLoadModuleData()) {
    loader = createLoadModuleData(getLoadModuleData(), Module);
  }
  return (...args) => Promise.resolve(loader(...args)).catch((error) => {
    console.error(`Error while attempting to call 'load' or 'loadModuleData' inside Holocron module ${Module.holocron.name}.`, error);
  });
};

// eslint-disable-next-line complexity
export function validateModule(Module, withConfig) {
  const config = {
    name: 'module',
    ...{ options: {} },
    ...withConfig || {},
    ...Module.holocron || {},
  };

  if (Module[REDUCER_KEY] && !config.reducer) {
    config.reducer = Module[REDUCER_KEY];
  } else if (config.reducer && !Module[REDUCER_KEY]) {
    // eslint-disable-next-line no-param-reassign
    Module[REDUCER_KEY] = config.reducer;
  }

  if (config.load && !Module[LOAD_KEY] && config.options.ssr) {
    // eslint-disable-next-line no-param-reassign
    Module[LOAD_KEY] = config.load;
  }

  if (Module.loadModuleData && !config.loadModuleData) {
    config.loadModuleData = Module.loadModuleData;
  }

  // eslint-disable-next-line no-param-reassign
  Module.holocron = {
    ...config,
  };

  return Module;
}

export function memoizeHolocronModule(Module, HolocronWrapper) {
  const HolocronModule = validateModule(Module);
  return hoistNonReactStatics(
    React.memo(
      hoistNonReactStatics(HolocronWrapper, HolocronModule),
      getShouldModuleReload(HolocronModule)
    ),
    HolocronWrapper,
    {
      displayName: getModuleDisplayName(getModuleName(HolocronModule)),
    }
  );
}
