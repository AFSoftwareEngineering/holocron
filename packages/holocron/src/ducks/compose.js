/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { loadModule } from './load';
import { getModuleLoadFn, getLoadModuleDataFn } from '../utility';

// This duck does not have a reducer
export default null;

export function composeModules(moduleConfigs, registry) {
  return (dispatch, getState, { fetchClient } = {}) => {
    const modulePromises = moduleConfigs.map((config) => {
      const { name } = config;
      return dispatch(loadModule(name, registry))
        .then((module) => {
          const load = getModuleLoadFn(module);

          if (load) {
            return dispatch(load(config.props));
          }

          const loadModuleData = getLoadModuleDataFn(module);

          if (loadModuleData) {
            return loadModuleData({
              store: { dispatch, getState },
              module,
              ownProps: config.props,
              fetchClient,
            });
          }

          return Promise.resolve();
        })
        .catch((error) => error);
    });
    return Promise.all(modulePromises);
  };
}
