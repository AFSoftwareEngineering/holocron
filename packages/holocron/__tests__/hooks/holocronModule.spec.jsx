/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/prop-types, import/no-extraneous-dependencies  */
import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { combineReducers as immutableCombineReducers } from 'redux-immutable';
import { Provider, useSelector } from 'react-redux';
import { fromJS } from 'immutable';
import { registerModule, getModule } from '../../src/moduleRegistry';

import { render } from '../__util__/renderer';

import { REDUCER_KEY, LOAD_KEY, MODULES_STORE_KEY } from '../../src/ducks/constants';

import createHolocronModule from '../../src/createHolocronModule';

// const sleep = (ms) => new Promise((resolve) => {
//   setTimeout(resolve, ms);
// });

// const warn = jest.spyOn(console, 'warn');
// const error = jest.spyOn(console, 'error');

describe(createHolocronModule.name, () => {
  const moduleName = 'my-module';

  describe('defaults', () => {
    const initialState = fromJS({
      modules: { [moduleName]: { foo: 'bar' } },
      holocron: {
        withReducers: [moduleName],
        loaded: [moduleName],
        failed: {},
        loading: {},
      },
    });

    test('creates holocron module and renders', () => {
      const Module = createHolocronModule(({ moduleLoadStatus, moduleState, dispatch }) => {
        expect(dispatch).toEqual(expect.any(Function));
        expect(moduleLoadStatus).toEqual(expect.any(String));
        expect(moduleState).toBeDefined();
        return <p>Hello!</p>;
      }, { name: moduleName });

      const result = render(<Module />);
      expect(result.toTree()).toMatchSnapshot();
      expect(result.toJSON()).toMatchSnapshot();
    });

    test('registers module', () => {
      const TestModule = ({ moduleLoadStatus }) => {
        const state = useSelector((_state) => _state.getIn([MODULES_STORE_KEY, moduleName]));
        // console.log(moduleLoadStatus, state.toJS());
        return <p>Hello!</p>;
      };

      TestModule[REDUCER_KEY] = (state) => state;

      registerModule(moduleName, TestModule);

      const Module = getModule(moduleName);

      const result = render(<Module />, { initialState });
      expect(result.toTree()).toMatchSnapshot();
      expect(result.toJSON()).toMatchSnapshot();
    });
  });

  describe('browser', () => {
    beforeAll(() => {
      global.BROWSER = true;
    });

    test('registers module', () => {
      const initialState = fromJS({
        modules: { [moduleName]: { foo: 'bar' } },
        holocron: {
          withReducers: [moduleName],
          loaded: [moduleName],
          failed: {},
          loading: {},
        },
      });

      const TestModule = ({ moduleLoadStatus }) => {
        const state = useSelector((_state) => _state.getIn([MODULES_STORE_KEY, moduleName]));
        // console.log(moduleLoadStatus, state.toJS());
        return <p>Hello!</p>;
      };

      TestModule[REDUCER_KEY] = (state) => state;

      registerModule(moduleName, TestModule);

      const Module = getModule(moduleName);

      const result = render(<Module />, { initialState });
      expect(result.toTree()).toMatchSnapshot();
      expect(result.toJSON()).toMatchSnapshot();
    });
  });
});
