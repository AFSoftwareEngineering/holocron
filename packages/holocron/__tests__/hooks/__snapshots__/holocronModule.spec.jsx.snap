// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHolocronModule browser registers module 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <Memo(HolocronModule(TestModule)) />,
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "rebuildReducer": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "load": undefined,
      "loadModuleData": undefined,
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "dispatch": [Function],
        "holocronLoadedStatus": false,
        "load": undefined,
        "loadModuleData": undefined,
        "moduleLoadStatus": "initial",
        "moduleState": null,
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": "Hello!",
        },
        "rendered": Array [
          "Hello!",
        ],
        "type": "p",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;

exports[`createHolocronModule browser registers module 2`] = `
<p>
  Hello!
</p>
`;

exports[`createHolocronModule creates holocron module and renders 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <Memo(HolocronModule(my-module)) />,
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "modules": Immutable.Map {},
      "rebuildReducer": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "load": undefined,
      "loadModuleData": undefined,
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "dispatch": [Function],
        "holocronLoadedStatus": false,
        "load": undefined,
        "loadModuleData": undefined,
        "moduleLoadStatus": "initial",
        "moduleState": null,
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": "Hello!",
        },
        "rendered": Array [
          "Hello!",
        ],
        "type": "p",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;

exports[`createHolocronModule creates holocron module and renders 2`] = `
<p>
  Hello!
</p>
`;

exports[`createHolocronModule defaults creates holocron module and renders 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <Memo(HolocronModule()) />,
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "modules": Immutable.Map {},
      "rebuildReducer": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "load": undefined,
      "loadModuleData": undefined,
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "dispatch": [Function],
        "holocronLoadedStatus": false,
        "load": undefined,
        "loadModuleData": undefined,
        "moduleLoadStatus": "initial",
        "moduleState": null,
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": "Hello!",
        },
        "rendered": Array [
          "Hello!",
        ],
        "type": "p",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;

exports[`createHolocronModule defaults creates holocron module and renders 2`] = `
<p>
  Hello!
</p>
`;

exports[`createHolocronModule defaults registers module 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <Memo(HolocronModule(TestModule)) />,
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "modules": Immutable.Map {
        "my-module": Object {
          "$$typeof": Symbol(react.memo),
          "@@holocron-module-reducer": [Function],
          "compare": null,
          "type": [Function],
        },
      },
      "rebuildReducer": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "load": undefined,
      "loadModuleData": undefined,
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "dispatch": [Function],
        "holocronLoadedStatus": false,
        "load": undefined,
        "loadModuleData": undefined,
        "moduleLoadStatus": "loaded",
        "moduleState": null,
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": "Hello!",
        },
        "rendered": Array [
          "Hello!",
        ],
        "type": "p",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;

exports[`createHolocronModule defaults registers module 2`] = `
<p>
  Hello!
</p>
`;

exports[`createHolocronModule registers module 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": <Memo(HolocronModule(TestModule)) />,
    "store": Object {
      "dispatch": [Function],
      "getState": [Function],
      "modules": Immutable.Map {
        "my-module": Object {
          "$$typeof": Symbol(react.memo),
          "@@holocron-module-reducer": [Function],
          "compare": null,
          "type": [Function],
        },
      },
      "rebuildReducer": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "load": undefined,
      "loadModuleData": undefined,
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "dispatch": [Function],
        "holocronLoadedStatus": false,
        "load": undefined,
        "loadModuleData": undefined,
        "moduleLoadStatus": "loaded",
        "moduleState": null,
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": "Hello!",
        },
        "rendered": Array [
          "Hello!",
        ],
        "type": "p",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;

exports[`createHolocronModule registers module 2`] = `
<p>
  Hello!
</p>
`;
